### PR TITLE
[FLINK-19423][jdbc] Fix ArrayIndexOutOfBoundsException when executing DELETE statement in JDBC upsert sink

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Fix ArrayIndexOutOfBoundsException when executing DELETE statement in JDBC upsert sink

## Brief change log

- Fix `JdbcDynamicOutputFormatBuilder#createDeleteExecutor`

## Verifying this change

- Added a test which can reproduce this problem.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @leonardBang

**Comment by wuchong** [comment]
Build is passed in my Azure: https://dev.azure.com/imjark/Flink/_build/results?buildId=276&view=results

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicOutputFormatBuilder.java [hunk 1]
```diff
@@ -182,7 +168,12 @@ public JdbcDynamicOutputFormatBuilder setFieldDataTypes(DataType[] fieldDataType
 		checkArgument(dmlOptions.getKeyFields().isPresent());
 		String[] pkNames = Arrays.stream(pkFields).mapToObj(k -> dmlOptions.getFieldNames()[k]).toArray(String[]::new);
 		String deleteSql = dmlOptions.getDialect().getDeleteStatement(dmlOptions.getTableName(), pkNames);
-		return createKeyedRowExecutor(dmlOptions.getDialect(), pkFields, pkTypes, deleteSql, fieldTypes);
+		JdbcRowConverter rowConverter = dmlOptions.getDialect().getRowConverter(RowType.of(pkTypes));
+		Function<RowData, RowData> keyExtractor = createRowKeyExtractor(fieldTypes, pkFields);
+		return JdbcBatchStatementExecutor.keyed(
+			deleteSql,
+			keyExtractor,
+			(st, key) -> rowConverter.toExternal(key, st));
```
Comment (leonardBang) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicOutputFormatBuilder.java:176]: nice catch

**Review by leonardBang — APPROVED** [review]
Thanks @wuchong for the contribution, the fix LGTM

**Comment by wuchong** [comment]
Merging...

**Pull request merged**